crypto/blake2/blake2s.c

```diff
@@ -0,0 +1,221 @@
+    P->key_length    = 0;
+    P->fanout        = 1;
+    P->depth         = 1;
+    store32(&P->leaf_length, 0);
+    store48(&P->node_offset, 0);
+    P->node_depth    = 0;
+    P->inner_length  = 0;
+    /* memset(P->reserved, 0, sizeof(P->reserved)); */
+    memset(P->salt,     0, sizeof(P->salt));
+    memset(P->personal, 0, sizeof(P->personal));
+    blake2s_init_param(c, P);
+    return 1;
+}
+
+/* Permute the state while xoring in the block of data. */
+static void blake2s_compress(BLAKE2S_CTX *S,
+                            const uint8_t block[BLAKE2S_BLOCKBYTES])
+{
+    uint32_t m[16];
+    uint32_t v[16];
+    size_t i;
+
+    for(i = 0; i < 16; ++i) {
+        m[i] = load32(block + i * sizeof(m[i]));
+    }
+
+    for(i = 0; i < 8; ++i) {
+        v[i] = S->h[i];
+    }
+
+    v[ 8] = blake2s_IV[0];
+    v[ 9] = blake2s_IV[1];
+    v[10] = blake2s_IV[2];
+    v[11] = blake2s_IV[3];
+    v[12] = S->t[0] ^ blake2s_IV[4];
+    v[13] = S->t[1] ^ blake2s_IV[5];
+    v[14] = S->f[0] ^ blake2s_IV[6];
+    v[15] = S->f[1] ^ blake2s_IV[7];
+#define G(r,i,a,b,c,d) \
+    do { \
+        a = a + b + m[blake2s_sigma[r][2*i+0]]; \
+        d = rotr32(d ^ a, 16); \
+        c = c + d; \
+        b = rotr32(b ^ c, 12); \
+        a = a + b + m[blake2s_sigma[r][2*i+1]]; \
+        d = rotr32(d ^ a, 8); \
+        c = c + d; \
+        b = rotr32(b ^ c, 7); \
+    } while(0)
+#define ROUND(r)  \
+    do { \
+        G(r,0,v[ 0],v[ 4],v[ 8],v[12]); \
+        G(r,1,v[ 1],v[ 5],v[ 9],v[13]); \
+        G(r,2,v[ 2],v[ 6],v[10],v[14]); \
+        G(r,3,v[ 3],v[ 7],v[11],v[15]); \
+        G(r,4,v[ 0],v[ 5],v[10],v[15]); \
+        G(r,5,v[ 1],v[ 6],v[11],v[12]); \
+        G(r,6,v[ 2],v[ 7],v[ 8],v[13]); \
+        G(r,7,v[ 3],v[ 4],v[ 9],v[14]); \
+    } while(0)
+    ROUND(0);
+    ROUND(1);
+    ROUND(2);
+    ROUND(3);
+    ROUND(4);
+    ROUND(5);
+    ROUND(6);
+    ROUND(7);
+    ROUND(8);
+    ROUND(9);
+
+    for(i = 0; i < 8; ++i) {
+        S->h[i] = S->h[i] ^ v[i] ^ v[i + 8];
+    }
+
+#undef G
+#undef ROUND
+}
+
+/* Absorb the input data into the hash state.  Always returns 1. */
+int BLAKE2s_Update(BLAKE2S_CTX *c, const void *data, size_t datalen)
+{
+    const uint8_t *in = data;
+    size_t fill;
+
+    while(datalen > 0) {
+        fill = sizeof(c->buf) - c->buflen;
+        /* Must be >, not >=, so that last block can be hashed differently */
+        if(datalen > fill) {
+            memcpy(c->buf + c->buflen, in, fill); /* Fill buffer */
+            blake2s_increment_counter(c, BLAKE2S_BLOCKBYTES);
+            blake2s_compress(c, c->buf); /* Compress */
+            c->buflen = 0;
+            in += fill;
+            datalen -= fill;
+        } else { /* datalen <= fill */
+            memcpy(c->buf + c->buflen, in, datalen);
+            c->buflen += datalen; /* Be lazy, do not compress */
+            return 1;
+        }
+    }
+
+    return 1;
+}
+
+/*
+ * Finalize the hash state in a way that avoids length extension attacks.
```

Comment:
Then how about something like

```
/*
 * The last compression call of BLAKE2 is different by design,
 * in order to avoid length extension attacks.
 */

(Make it clear that the comment is about the design and not our implementation.)